### ISSUE TITLE
yield called out of block

### ISSUE THREAD

**Issue by msoltyspl** [logstash-plugins/logstash-filter-ruby#41]
Versions tested: 6.1.3, 6.2.1
Operating system: linux (ubuntu xenial)

We have run into issues when using ruby filter with 'path' option and when creating events (via event.clone). From what we can see in the filter's code - when using 'path' - the code itself distinguishes between the new (to be yielded) events and the original one. But in practice - whenever we use event.clone, we get:

```
[2018-02-13T10:56:41,377][ERROR][logstash.pipeline        ] Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {:pipeline_id=>"main", "exception"=>"yield called out of block", "backtrace"=>["/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-filter-ruby-3.1.3/lib/logstash/filters/ruby.rb:116:in `block in file_script'", "org/jruby/RubyArray.java:1734:in `each'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-filter-ruby-3.1.3/lib/logstash/filters/ruby.rb:111:in `file_script'", "/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-filter-ruby-3.1.3/lib/logstash/filters/ruby.rb:84:in `filter'", "/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:145:in `do_filter'", "/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:164:in `block in multi_filter'", "org/jruby/RubyArray.java:1734:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:161:in `multi_filter'", "/usr/share/logstash/logstash-core/lib/logstash/filter_delegator.rb:47:in `multi_filter'", "(eval):248:in `block in initialize'", "org/jruby/RubyArray.java:1734:in `each'", "(eval):239:in `block in initialize'", "(eval):221:in `block in filter_func'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:447:in `filter_batch'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:426:in `worker_loop'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:385:in `block in start_workers'"], :thread=>"#<Thread:0x3ff2bfd1@/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:246 sleep>"}
```

This can be recreated with trivial:

```
def register(params)
end
def filter(event)
    e = event.clone
    event.cancel
    return [e]
end
```

Called by e.g.:
```                ruby {
                        id => "f_split_upstream"
                        path => '/usr/share/logstash-plugins/split_upstream/split_upstream.rb'
                }
```

Imporant things to note: any tests defined in the file **do** work correctly. We have to pass actual events after the logstash is started for it to bail out with exception.

**Comment by guyboertje** [comment]
The recent [blog post by Joao Duarte](https://www.elastic.co/blog/category/engineering) does similar stuff without the clone.


Investigating...

**Comment by Crack** [comment]
```ruby
  def filter(event, &block)
    if @code
      inline_script(event, &block)
    elsif @path
      file_script(event)
    end
end
```
Just a guess, but shouldn't `file_script` be called with `&block`?

**Comment by Crack** [comment]
I run this change (`file_script(event, &block)`) though my local logstash test, and ruby filter was able to properly output multiple events.

**Comment by zmays** [comment]
I'm having the same issue when trying to use Logstash::Event.new() and returning that event (also using path to specify a ruby script): 

`def filter(event)
  e = LogStash::Event.new()
  e.set("test", "test")
  return [e] # return empty array to cancel event
end
`

**Comment by msoltyspl** [comment]
Just to note - the workaround proposed by @Crack solves the issue (adding missing &block in invocation).

**Comment by palashkulsh** [comment]
for anyone facing "yield called out of block" error when passing multiple events in logstash from ruby file filter function , then simply update the logstash-filter-ruby plugin by running 

> ./bin/logstash-plugin update logstash-filter-ruby
it worked for me.

**Comment by jsvd** [comment]
right, thanks @palashkulsh. the fix was implemented in version [3.1.4](https://github.com/logstash-plugins/logstash-filter-ruby/blob/master/CHANGELOG.md#314).
